### PR TITLE
Warn against custom non-lowercase attributes

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/react/issues/10590.

### CODE REVIEW

File: src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js [hunk 1]
```diff
@@ -133,5 +133,25 @@ describe('ReactDOM unknown attribute', () => {
...
+      expect(el.firstChild.getAttribute('helloworld')).toBe('something');
+
+      expectDev(console.error.calls.count()).toBe(1);
+      expectDev(
+        normalizeCodeLocInfo(console.error.calls.argsFor(0)[0]),
+      ).toMatch(
+        'React does not recognize the `helloWorld` prop on a DOM element. ' +
+          'If you intentionally wanted it to appear in the DOM as a custom ' +
+          'attribute, spell it as the lowercase `helloworld` instead. ' +
+          'If you passed it accidentally from a parent component, remove ' +
```
Comment (gaearon) [inline src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js]: I think it’s important to mention this since people will often get this warning due to spreading.

File: src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js [hunk 2]
```diff
@@ -909,25 +909,25 @@ describe('ReactDOMServerIntegration', () => {
       });
 
       itRenders('unknown data- attributes with casing', async render => {
-        const e = await render(<div data-fooBar="true" />);
-        expect(e.getAttribute('data-fooBar')).toBe('true');
+        const e = await render(<div data-fooBar="true" />, 1);
+        expect(e.getAttribute('data-foobar')).toBe('true');
```
Comment (gaearon) [inline src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js]: Changing these calls wasn't essential but I figured I'd emphasize they are normalized by the browser anyway.

File: src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js [hunk 3]
```diff
@@ -909,25 +909,25 @@ describe('ReactDOMServerIntegration', () => {
...
       });
 
       itRenders('unknown data- attributes with boolean true', async render => {
-        const e = await render(<div data-fooBar={true} />);
-        expect(e.getAttribute('data-fooBar')).toBe('true');
+        const e = await render(<div data-foobar={true} />);
+        expect(e.getAttribute('data-foobar')).toBe('true');
```
Comment (gaearon) [inline src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js]: These tests were not specifically about casing so I fixed them.

File: src/renderers/dom/shared/hooks/ReactDOMUnknownPropertyHook.js [hunk 4]
```diff
@@ -39,7 +39,10 @@ if (__DEV__) {
   var warnedProperties = {};
   var hasOwnProperty = Object.prototype.hasOwnProperty;
   var EVENT_NAME_REGEX = /^on[A-Z]/;
-  var ARIA_NAME_REGEX = /^aria-/i;
+  var rARIA = new RegExp('^(aria)-[' + DOMProperty.ATTRIBUTE_NAME_CHAR + ']*$');
+  var rARIACamel = new RegExp(
+    '^(aria)[A-Z][' + DOMProperty.ATTRIBUTE_NAME_CHAR + ']*$',
+  );
```
Comment (gaearon) [inline src/renderers/dom/shared/hooks/ReactDOMUnknownPropertyHook.js]: Copy pasta from ARIA hook.

File: src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js (hunk 1 above)
Comment (jquense) [reply src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js]: It seems like practically this is going to make the unknown DOM warning permanent (modulo some cases). Granted it's probably not what the user intends it seems a bit heavy handed.

Comment (jquense) [reply src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js]: or perhaps maybe just that it's going to give the impression that everything is fine when they aren't getting this warning, while `foo="[Object object]"` still gets passed

Comment (nhunzaker) [reply src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js]: The react prop name can still have casing, as long as it's passed as lowercase to the HTML element. This doesn't allow spreading of attributes, but you can do it at least.

I think it's important to communicate what this is protecting against. I don't know that this should live in the warning, but I do agree that a developer not familiar with the problem-space could see this as an arbitrary restriction. 

We could turn it into a "React's got your back" kind of thing if we effectively communicate it.

File: fixtures/attribute-behavior/AttributeTableSnapshot.md [hunk 5]
```diff
@@ -701,23 +701,23 @@
 ## `as` (on `<div>`)
 | Test Case | Flags | Result |
 | --- | --- | --- |
-| `as=(string)`| (changed)| `"a string"` |
+| `as=(string)`| (initial)| `<empty string>` |
```
Comment (gaearon) [inline fixtures/attribute-behavior/AttributeTableSnapshot.md]: This is an unrelated change, and seems like Chrome version difference. Setting `<link as="whatever">` doesn't return `"whatever"` from `link.as` property anymore. But it works with valid values like `"audio"`.

File: src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js (hunk 1 above)
Comment (jquense) [reply src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js]: What about custom elements (or does that not hit here)? Chance of a false positive? If we are trying to get out of folks way about passing what they want to the DOM this feels a bit arbitrary as you say.

**Comment by gaearon** [comment]
I agree but it’s also a bit hard to guess in advance what kind of things people would be using this for. So I’d wait until 16 and then look at these warnings again and maybe unify them more.

File: fixtures/attribute-behavior/AttributeTableSnapshot.md [hunk 6]
```diff
@@ -9276,27 +9276,27 @@
...
-| `selectedIndex=(string)`| (initial)| `<number: -1>` |
-| `selectedIndex=(empty string)`| (initial)| `<number: -1>` |
-| `selectedIndex=(array with string)`| (initial)| `<number: -1>` |
-| `selectedIndex=(empty array)`| (initial)| `<number: -1>` |
-| `selectedIndex=(object)`| (initial)| `<number: -1>` |
-| `selectedIndex=(numeric string)`| (initial)| `<number: -1>` |
-| `selectedIndex=(-1)`| (initial)| `<number: -1>` |
-| `selectedIndex=(0)`| (initial)| `<number: -1>` |
-| `selectedIndex=(integer)`| (initial)| `<number: -1>` |
+| `selectedIndex=(string)`| (initial, warning)| `<number: -1>` |
```
Comment (sebmarkbage) [inline fixtures/attribute-behavior/AttributeTableSnapshot.md]: I love that we finally got a nice warning on these for free.

Comment (sebmarkbage) [reply fixtures/attribute-behavior/AttributeTableSnapshot.md]: Although I guess this will just make people try `selectedindex` instead. 😕

File: src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js (hunk 1 above)
Comment (jquense) [reply src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js]: sigh didn't see the original issue, sorry for the noise :P
